app/models/user.rb

```diff
@@ -16,6 +16,11 @@ class User < ApplicationRecord
   has_many :projects, dependent: :destroy
   has_many :user_providers, dependent: :destroy
 
+  def progress_for(course)
+    @progress ||= Hash.new { |h, c| h[c] = CourseProgress.new(c, self) }
```

Comment:
It is quite common that we need to get the `CourseProgress` object for the same user/course more than once in the same request, so this caches it.